// @subsquid/fuel-data/lib/raw-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blocks = exports.Block = exports.Transaction = exports.UpgradePurpose = exports.Policies = exports.TransactionStatus = exports.Receipt = exports.TransactionOutput = exports.TransactionInput = exports.GetBlockHeader = exports.BlockHeader = exports.GetBlockHash = exports.LatestBlockHeight = void 0;
const util_internal_validation_1 = require("@subsquid/util-internal-validation");
exports.LatestBlockHeight = (0, util_internal_validation_1.object)({
    chain: (0, util_internal_validation_1.object)({
        latestBlock: (0, util_internal_validation_1.object)({
            header: (0, util_internal_validation_1.object)({
                height: util_internal_validation_1.STRING_NAT,
            })
        })
    })
});
exports.GetBlockHash = (0, util_internal_validation_1.object)({
    block: (0, util_internal_validation_1.option)((0, util_internal_validation_1.object)({
        id: util_internal_validation_1.STRING
    }))
});
exports.BlockHeader = (0, util_internal_validation_1.object)({
    id: util_internal_validation_1.BYTES,
    height: util_internal_validation_1.STRING_NAT,
    daHeight: util_internal_validation_1.BIG_NAT,
    transactionsRoot: util_internal_validation_1.BYTES,
    transactionsCount: util_internal_validation_1.STRING_NAT,
    messageReceiptCount: util_internal_validation_1.STRING_NAT,
    prevRoot: util_internal_validation_1.BYTES,
    time: util_internal_validation_1.BIG_NAT,
    applicationHash: util_internal_validation_1.BYTES,
    consensusParametersVersion: util_internal_validation_1.STRING_NAT,
    stateTransitionBytecodeVersion: util_internal_validation_1.STRING_NAT,
    eventInboxRoot: util_internal_validation_1.BYTES,
    messageOutboxRoot: util_internal_validation_1.BYTES,
});
exports.GetBlockHeader = (0, util_internal_validation_1.object)({
    block: (0, util_internal_validation_1.option)((0, util_internal_validation_1.object)({
        header: exports.BlockHeader
    }))
});
exports.TransactionInput = (0, util_internal_validation_1.taggedUnion)('__typename', {
    InputCoin: (0, util_internal_validation_1.object)({
        utxoId: util_internal_validation_1.BYTES,
        owner: util_internal_validation_1.BYTES,
        amount: util_internal_validation_1.BIG_NAT,
        assetId: util_internal_validation_1.BYTES,
        txPointer: util_internal_validation_1.STRING,
        witnessIndex: util_internal_validation_1.INT,
        predicateGasUsed: util_internal_validation_1.BIG_NAT,
        predicate: util_internal_validation_1.BYTES,
        predicateData: util_internal_validation_1.BYTES,
    }),
    InputContract: (0, util_internal_validation_1.object)({
        utxoId: util_internal_validation_1.BYTES,
        balanceRoot: util_internal_validation_1.BYTES,
        stateRoot: util_internal_validation_1.BYTES,
        txPointer: util_internal_validation_1.STRING,
        contractId: util_internal_validation_1.BYTES,
    }),
    InputMessage: (0, util_internal_validation_1.object)({
        sender: util_internal_validation_1.BYTES,
        recipient: util_internal_validation_1.BYTES,
        amount: util_internal_validation_1.BIG_NAT,
        nonce: util_internal_validation_1.BYTES,
        witnessIndex: util_internal_validation_1.STRING_NAT,
        predicateGasUsed: util_internal_validation_1.BIG_NAT,
        data: util_internal_validation_1.BYTES,
        predicate: util_internal_validation_1.BYTES,
        predicateData: util_internal_validation_1.BYTES,
    }),
});
exports.TransactionOutput = (0, util_internal_validation_1.taggedUnion)('__typename', {
    CoinOutput: (0, util_internal_validation_1.object)({
        to: util_internal_validation_1.BYTES,
        amount: util_internal_validation_1.BIG_NAT,
        assetId: util_internal_validation_1.BYTES,
    }),
    ContractOutput: (0, util_internal_validation_1.object)({
        inputIndex: util_internal_validation_1.STRING_NAT,
        balanceRoot: util_internal_validation_1.BYTES,
        stateRoot: util_internal_validation_1.BYTES,
    }),
    ChangeOutput: (0, util_internal_validation_1.object)({
        to: util_internal_validation_1.BYTES,
        amount: util_internal_validation_1.BIG_NAT,
        assetId: util_internal_validation_1.BYTES,
    }),
    VariableOutput: (0, util_internal_validation_1.object)({
        to: util_internal_validation_1.BYTES,
        amount: util_internal_validation_1.BIG_NAT,
        assetId: util_internal_validation_1.BYTES,
    }),
    ContractCreated: (0, util_internal_validation_1.object)({
        contract: util_internal_validation_1.BYTES,
        stateRoot: util_internal_validation_1.BYTES,
    }),
});
exports.Receipt = (0, util_internal_validation_1.object)({
    id: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    pc: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    is: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    to: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    toAddress: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    amount: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    assetId: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    gas: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    param1: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    param2: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    val: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    ptr: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    digest: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    reason: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    ra: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    rb: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    rc: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    rd: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    len: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    receiptType: (0, util_internal_validation_1.oneOf)({
        CALL: (0, util_internal_validation_1.constant)('CALL'),
        RETURN: (0, util_internal_validation_1.constant)('RETURN'),
        RETURN_DATA: (0, util_internal_validation_1.constant)('RETURN_DATA'),
        PANIC: (0, util_internal_validation_1.constant)('PANIC'),
        REVERT: (0, util_internal_validation_1.constant)('REVERT'),
        LOG: (0, util_internal_validation_1.constant)('LOG'),
        LOG_DATA: (0, util_internal_validation_1.constant)('LOG_DATA'),
        TRANSFER: (0, util_internal_validation_1.constant)('TRANSFER'),
        TRANSFER_OUT: (0, util_internal_validation_1.constant)('TRANSFER_OUT'),
        SCRIPT_RESULT: (0, util_internal_validation_1.constant)('SCRIPT_RESULT'),
        MESSAGE_OUT: (0, util_internal_validation_1.constant)('MESSAGE_OUT'),
        MINT: (0, util_internal_validation_1.constant)('MINT'),
        BURN: (0, util_internal_validation_1.constant)('BURN'),
    }),
    result: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    gasUsed: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    data: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    sender: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    recipient: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    nonce: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    contractId: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    subId: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
});
exports.TransactionStatus = (0, util_internal_validation_1.taggedUnion)('__typename', {
    SubmittedStatus: (0, util_internal_validation_1.object)({ time: util_internal_validation_1.BIG_NAT }),
    SuccessStatus: (0, util_internal_validation_1.object)({
        transactionId: util_internal_validation_1.BYTES,
        time: util_internal_validation_1.BIG_NAT,
        programState: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.object)({
            returnType: (0, util_internal_validation_1.oneOf)({
                return: (0, util_internal_validation_1.constant)('RETURN'),
                returnData: (0, util_internal_validation_1.constant)('RETURN_DATA'),
                revert: (0, util_internal_validation_1.constant)('REVERT'),
            }),
            data: util_internal_validation_1.BYTES,
        })),
        receipts: (0, util_internal_validation_1.option)((0, util_internal_validation_1.array)(exports.Receipt)),
        totalGas: util_internal_validation_1.BIG_NAT,
        totalFee: util_internal_validation_1.BIG_NAT,
    }),
    SqueezedOutStatus: (0, util_internal_validation_1.object)({ reason: util_internal_validation_1.STRING }),
    FailureStatus: (0, util_internal_validation_1.object)({
        transactionId: util_internal_validation_1.BYTES,
        time: util_internal_validation_1.BIG_NAT,
        reason: util_internal_validation_1.STRING,
        programState: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.object)({
            returnType: (0, util_internal_validation_1.oneOf)({
                RETURN: (0, util_internal_validation_1.constant)('RETURN'),
                RETURN_DATA: (0, util_internal_validation_1.constant)('RETURN_DATA'),
                REVERT: (0, util_internal_validation_1.constant)('REVERT'),
            }),
            data: util_internal_validation_1.BYTES,
        })),
        receipts: (0, util_internal_validation_1.option)((0, util_internal_validation_1.array)(exports.Receipt)),
        totalGas: util_internal_validation_1.BIG_NAT,
        totalFee: util_internal_validation_1.BIG_NAT,
    })
});
exports.Policies = (0, util_internal_validation_1.object)({
    tip: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    witnessLimit: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    maturity: (0, util_internal_validation_1.nullable)(util_internal_validation_1.STRING_NAT),
    maxFee: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
});
exports.UpgradePurpose = (0, util_internal_validation_1.taggedUnion)('__typename', {
    ConsensusParametersPurpose: (0, util_internal_validation_1.object)({
        witnessIndex: util_internal_validation_1.STRING_NAT,
        checksum: util_internal_validation_1.BYTES,
    }),
    StateTransitionPurpose: (0, util_internal_validation_1.object)({
        root: util_internal_validation_1.BYTES,
    }),
});
exports.Transaction = (0, util_internal_validation_1.object)({
    id: util_internal_validation_1.BYTES,
    inputAssetIds: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.array)(util_internal_validation_1.BYTES)),
    inputContracts: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.array)(util_internal_validation_1.BYTES)),
    inputContract: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.object)({
        utxoId: util_internal_validation_1.BYTES,
        balanceRoot: util_internal_validation_1.BYTES,
        stateRoot: util_internal_validation_1.BYTES,
        txPointer: util_internal_validation_1.STRING,
        contractId: util_internal_validation_1.BYTES
    })),
    policies: (0, util_internal_validation_1.nullable)(exports.Policies),
    scriptGasLimit: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    maturity: (0, util_internal_validation_1.nullable)(util_internal_validation_1.STRING_NAT),
    mintAmount: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    mintAssetId: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    mintGasPrice: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BIG_NAT),
    txPointer: (0, util_internal_validation_1.nullable)(util_internal_validation_1.STRING),
    isScript: util_internal_validation_1.BOOLEAN,
    isCreate: util_internal_validation_1.BOOLEAN,
    isMint: util_internal_validation_1.BOOLEAN,
    isUpgrade: util_internal_validation_1.BOOLEAN,
    isUpload: util_internal_validation_1.BOOLEAN,
    inputs: (0, util_internal_validation_1.option)((0, util_internal_validation_1.array)(exports.TransactionInput)),
    outputs: (0, util_internal_validation_1.option)((0, util_internal_validation_1.array)(exports.TransactionOutput)),
    outputContract: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.object)({
        inputIndex: util_internal_validation_1.STRING_NAT,
        balanceRoot: util_internal_validation_1.BYTES,
        stateRoot: util_internal_validation_1.BYTES,
    })),
    witnesses: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.array)(util_internal_validation_1.BYTES)),
    receiptsRoot: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    status: exports.TransactionStatus,
    script: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    scriptData: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    salt: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    storageSlots: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.array)(util_internal_validation_1.BYTES)),
    rawPayload: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    bytecodeWitnessIndex: (0, util_internal_validation_1.nullable)(util_internal_validation_1.STRING_NAT),
    bytecodeRoot: (0, util_internal_validation_1.nullable)(util_internal_validation_1.BYTES),
    subsectionIndex: (0, util_internal_validation_1.nullable)(util_internal_validation_1.STRING_NAT),
    subsectionsNumber: (0, util_internal_validation_1.nullable)(util_internal_validation_1.STRING_NAT),
    proofSet: (0, util_internal_validation_1.nullable)((0, util_internal_validation_1.array)(util_internal_validation_1.BYTES)),
    upgradePurpose: (0, util_internal_validation_1.nullable)(exports.UpgradePurpose),
});
exports.Block = (0, util_internal_validation_1.object)({
    header: exports.BlockHeader,
    transactions: (0, util_internal_validation_1.array)(exports.Transaction),
});
exports.Blocks = (0, util_internal_validation_1.object)({
    blocks: (0, util_internal_validation_1.object)({
        nodes: (0, util_internal_validation_1.array)(exports.Block),
    }),
});
