// @fuel-ts/versions/dist/cli.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/cli.ts
var cli_exports = {};
__export(cli_exports, {
  compareSystemVersions: () => compareSystemVersions,
  eitherOr: () => eitherOr,
  fuelUpLink: () => fuelUpLink,
  getBuiltinVersions: () => getBuiltinVersions,
  getSystemForc: () => getSystemForc,
  getSystemFuelCore: () => getSystemFuelCore,
  getSystemVersion: () => getSystemVersion,
  getSystemVersions: () => getSystemVersions,
  runVersions: () => runVersions
});
module.exports = __toCommonJS(cli_exports);
var import_chalk2 = __toESM(require("chalk"));
var import_cli_table = __toESM(require("cli-table"));

// src/lib/colorizeUserVersion.ts
var import_chalk = __toESM(require("chalk"));
var colorizeUserVersion = (params) => {
  const { version, isGt, isOk } = params;
  if (isGt) {
    return import_chalk.default.cyan(version);
  }
  if (isOk) {
    return import_chalk.default.green(version);
  }
  return import_chalk.default.red(version);
};

// src/lib/getBuiltinVersions.ts
function getBuiltinVersions() {
  return {
    FORC: "0.60.0",
    FUEL_CORE: "0.27.0",
    FUELS: "0.89.1"
  };
}

// src/lib/semver.ts
function parseVersion(version) {
  const [major, minor, patch] = version.split(".").map((v) => parseInt(v, 10));
  return { major, minor, patch };
}
function versionDiffs(version1, version2) {
  const semver1 = parseVersion(version1);
  const semver2 = parseVersion(version2);
  const major = semver1.major - semver2.major;
  const minor = semver1.minor - semver2.minor;
  const patch = semver1.patch - semver2.patch;
  return {
    major,
    minor,
    patch,
    fullVersionDiff: major || minor || patch
  };
}
function gt(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff > 0;
}
function eq(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff === 0;
}

// src/lib/compareSystemVersions.ts
function compareSystemVersions(params) {
  const { systemForcVersion, systemFuelCoreVersion } = params;
  const versions = getBuiltinVersions();
  const systemForcIsGt = gt(systemForcVersion, versions.FORC);
  const systemFuelCoreIsGt = gt(systemFuelCoreVersion, versions.FUEL_CORE);
  const systemForcIsEq = eq(systemForcVersion, versions.FORC);
  const systemFuelCoreIsEq = eq(systemFuelCoreVersion, versions.FUEL_CORE);
  const systemForcIsLt = !systemForcIsGt && !systemForcIsEq;
  const systemFuelCoreIsLt = !systemFuelCoreIsGt && !systemFuelCoreIsEq;
  return {
    systemForcIsGt,
    systemFuelCoreIsGt,
    systemForcIsEq,
    systemFuelCoreIsEq,
    systemForcIsLt,
    systemFuelCoreIsLt
  };
}

// src/lib/fuelUpLink.ts
var fuelUpLink = "https://github.com/fuellabs/fuelup";

// src/lib/getSystemVersions.ts
var import_child_process = require("child_process");
var versionReg = /[0-9]+\.[0-9]+\.[0-9]/;
var defaultForcCommand = "forc";
var defaultFuelCoreCommand = "fuel-core";
var getSystemVersion = (command) => {
  let version = null;
  let error = null;
  try {
    const contents = (0, import_child_process.execSync)(command, {
      stdio: ["pipe", "pipe", "ignore"],
      encoding: "utf8"
    }).toString();
    if (versionReg.test(contents)) {
      version = contents.match(versionReg)?.[0];
    } else {
      throw new Error(contents);
    }
  } catch (err) {
    error = err;
  }
  return {
    error,
    version
  };
};
function getSystemForc(forcPath = defaultForcCommand) {
  const { error, version: v } = getSystemVersion(`${forcPath} --version`);
  return { error, systemForcVersion: v, systemForcPath: forcPath };
}
function getSystemFuelCore(fuelCorePath = defaultFuelCoreCommand) {
  const { error, version: v } = getSystemVersion(`${fuelCorePath} --version`);
  return { error, systemFuelCoreVersion: v, systemFuelCorePath: fuelCorePath };
}
function getSystemVersions(params = {}) {
  const { forcPath, fuelCorePath } = params;
  const { error: errorForc, systemForcVersion, systemForcPath } = getSystemForc(forcPath);
  const {
    error: errorCore,
    systemFuelCoreVersion,
    systemFuelCorePath
  } = getSystemFuelCore(fuelCorePath);
  const error = errorForc ?? errorCore;
  return {
    error,
    systemForcVersion,
    systemForcPath,
    systemFuelCoreVersion,
    systemFuelCorePath
  };
}

// src/cli.ts
var eitherOr = (val1, val2) => val1 ?? val2;
function runVersions(params = {}) {
  const { error, info } = console;
  const supportedVersions = getBuiltinVersions();
  const cliTable = new import_cli_table.default({
    head: ["", import_chalk2.default.bold("Supported"), import_chalk2.default.bold(`Yours / System`), import_chalk2.default.bold("System Path")]
  });
  const {
    error: systemError,
    systemForcVersion,
    systemFuelCoreVersion,
    systemForcPath,
    systemFuelCorePath
  } = getSystemVersions(params);
  const comparisons = compareSystemVersions({
    systemForcVersion: eitherOr(systemForcVersion, "0"),
    systemFuelCoreVersion: eitherOr(systemFuelCoreVersion, "0")
  });
  const userForcColorized = colorizeUserVersion({
    version: eitherOr(systemForcVersion, "\u2014"),
    isGt: comparisons.systemForcIsGt,
    isOk: comparisons.systemForcIsEq
  });
  const userFuelCoreColorized = colorizeUserVersion({
    version: eitherOr(systemFuelCoreVersion, "\u2014"),
    isGt: comparisons.systemFuelCoreIsGt,
    isOk: comparisons.systemFuelCoreIsEq
  });
  cliTable.push(["Forc", supportedVersions.FORC, userForcColorized, systemForcPath]);
  cliTable.push([
    "Fuel-Core",
    supportedVersions.FUEL_CORE,
    userFuelCoreColorized,
    systemFuelCorePath
  ]);
  const someIsGt = comparisons.systemForcIsGt || comparisons.systemFuelCoreIsGt;
  const bothAreExact = comparisons.systemForcIsEq && comparisons.systemFuelCoreIsEq;
  let exitCode;
  if (someIsGt) {
    exitCode = 0;
    info(cliTable.toString());
    info(`
Your system's components are newer than the ones supported!`);
  } else if (bothAreExact) {
    exitCode = 0;
    info(cliTable.toString());
    info(`
You have all the right versions! \u26A1`);
  } else if (systemError) {
    exitCode = 1;
    error(cliTable.toString());
    error("\n - Make sure you have Forc and Fuel-Core installed");
    error("   >> Error: ", systemError.message);
  } else {
    exitCode = 1;
    error(cliTable.toString());
    error(`
 - You're using outdated versions`);
  }
  if (exitCode === 1) {
    error(`  ${import_chalk2.default.green(fuelUpLink)}`);
  }
  process.exit(exitCode);
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  compareSystemVersions,
  eitherOr,
  fuelUpLink,
  getBuiltinVersions,
  getSystemForc,
  getSystemFuelCore,
  getSystemVersion,
  getSystemVersions,
  runVersions
});
